### PR TITLE
Enforcing usage of Key Vaults in Add/Update monitoring virtual machine workbooks.

### PR DESCRIPTION
**Pull request by bhsu-ms**
## PR Checklist
Enforcing usage of Key Vaults in monitoring vm config related workbooks due to security concerns. With this update, users cannot add/update monitoring vm config without giving access to Key Vaults. Added an informational banner asking user to reference passwords from Key Vaults and two error banners when access is not provided.
* [x] Explain your changes, so people looking at the PR know *what* and *why*, the code changes are the *how*.

### If adding or updating templates:
* [x] post a screenshot of templates and/or gallery changes
* [x] ensure your template has a corresponding gallery entry in the gallery folder
* [x] grep `/subscription/` and ensure that your parameters don't have any hardcoded resourceIds __or they will fail PR validation__
* [x] remove `fallbackResourceIds` and `fromTemplateId` fields from your template workbook __or they will fail PR validation__
image
image
link to workbooks: https://ms.portal.azure.com/?feature.includePreviewTemplates=true&feature.localtemplates=true&feature.workloadinsights=true&feature.workbookGalleryRedirect=https%3A%2F%2Freadablesecondaries.blob.core.windows.net%2Fazure-monitor-workbook-templates%2Fpackage#blade/Microsoft_Azure_Monitoring/AzureMonitoringBrowseBlade/sqlWorkloadInsights

### CODE REVIEW

**Review by masree — APPROVED** [review]

File: Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook [hunk 1]
```diff
@@ -502,7 +510,7 @@
           {
             "type": 1,
             "content": {
-              "json": "#### Enable access to Azure Key Vaults (Optional)\nIf your connection strings reference secrets from Key Vaults, you need to provide access to it from the virtual machine via a Managed Service Identity."
+              "json": "#### Enable access to Azure Key Vaults\nPlease provide access to the virtual machine from the Key Vaults referenced in the connection strings via Managed Service Identity."
```
Comment (dimitri-furman) [inline Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook:513]: proposed message, for clarity: `#### Enable access to Azure Key Vault\nUse Key Vault access policy to provide the Get Secret permission for all key vaults referenced in configuration. Permission must be provided to the Managed Service Identity of the monitoring virtual machine.`

File: Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook [hunk 2]
```diff
@@ -705,12 +713,52 @@
...
+                "parameterName": "KeyVaults",
+                "comparison": "isNotEqualTo"
+              }
+            ],
+            "name": "links - 5"
+          },
+          {
+            "type": 1,
+            "content": {
+              "json": "Please provide access to the Key Vaults",
```
Comment (dimitri-furman) [inline Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook:736]: `Please provide access to Key Vault`

File: Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook [hunk 3]
```diff
@@ -705,12 +713,52 @@
...
+              "parameterName": "KeyVaultSubscriptions",
+              "comparison": "isEqualTo"
             },
-            "name": "links - 5"
+            "name": "text - 7"
+          },
+          {
+            "type": 1,
+            "content": {
+              "json": "Please provide access to the Key Vaults",
```
Comment (dimitri-furman) [inline Workbooks/Workloads/SQL/Add monitoring virtual machine/Add monitoring virtual machine.workbook:748]: `Please provide access to Key Vault`

File: Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook [hunk 4]
```diff
@@ -150,6 +150,14 @@
       },
       "name": "Saved config not found message"
     },
+    {
+      "type": 1,
+      "content": {
+        "json": "Please reference passwords from Key Vaults.",
```
Comment (dimitri-furman) [inline Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook:156]: `Please reference passwords stored as Key Vault secrets.`

File: Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook [hunk 5]
```diff
@@ -280,7 +288,7 @@
     {
       "type": 1,
       "content": {
-        "json": "#### Enable access to Azure Key Vaults (Optional)\nIf your connection strings reference secrets from Key Vaults, you need to provide access to it from the virtual machine via a Managed Service Identity."
+        "json": "#### Enable access to Azure Key Vaults\nPlease provide access to the virtual machine from the Key Vaults referenced in the connection strings via Managed Service Identity."
```
Comment (dimitri-furman) [inline Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook:291]: `#### Enable access to Azure Key Vault\nUse Key Vault access policy to provide the Get Secret permission for all key vaults referenced in configuration. Permission must be provided to the Managed Service Identity of the monitoring virtual machine.`

File: Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook [hunk 6]
```diff
@@ -493,12 +501,52 @@
...
+          "parameterName": "KeyVaults",
+          "comparison": "isNotEqualTo"
+        }
+      ],
+      "name": "Valid Form"
+    },
+    {
+      "type": 1,
+      "content": {
+        "json": "Please provide access to the Key Vaults",
```
Comment (dimitri-furman) [inline Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook:524]: `Please provide access to Key Vault`

File: Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook [hunk 7]
```diff
@@ -493,12 +501,52 @@
...
+        "parameterName": "KeyVaultSubscriptions",
+        "comparison": "isEqualTo"
       },
-      "name": "Valid Form"
+      "name": "text - 14"
+    },
+    {
+      "type": 1,
+      "content": {
+        "json": "Please provide access to the Key Vaults",
```
Comment (dimitri-furman) [inline Workbooks/Workloads/Update monitoring vm config/Update monitoring vm config.workbook:536]: `Please provide access to Key Vault`

**Review by dimitri-furman — CHANGES_REQUESTED** [review]
Suggested tweaks to user-visible messages

**Review by gardnerjr — COMMENTED** [review]
did you explicitly request review from the workbooks team or is it because of CODEOWNERS says we own the files (the default state?)  If you own these files, you should add your github team to the CODEOWNERS file for this path, so that it doesn't require workbooks team to review your content.

**Comment by gardnerjr** [comment]
for the code owners part, it might just be a syntax issue in the `CODEOWNERS` file that's not catching these.
codeowners has:

```
/Workbooks/Workloads/SQL** @microsoft/sqlexternalmonitoring
```

but it looks like it is missing a trailing `/` to mark it as "Everything inside SQL folder" instead of "every file with SQL in the name" like it is now.

looks like it should be:
```
/Workbooks/Workloads/SQL/** @microsoft/sqlexternalmonitoring
```

**Review by dimitri-furman — APPROVED** [review]

**Comment by bhsu-ms** [comment]
> for the code owners part, it might just be a syntax issue in the `CODEOWNERS` file that's not catching these. codeowners has:
> 
> ```
> /Workbooks/Workloads/SQL** @microsoft/sqlexternalmonitoring
> ```
> 
> but it looks like it is missing a trailing `/` to mark it as "Everything inside SQL folder" instead of "every file with SQL in the name" like it is now.
> 
> looks like it should be:
> 
> ```
> /Workbooks/Workloads/SQL/** @microsoft/sqlexternalmonitoring
> ```

Hi John, thanks for spotting that.

I have updated the CODEOWNERS file accordingly.

**Review by gardnerjr — APPROVED** [review]
haven't really  validated *content*, but have looked at codeowners and general

**Pull request merged**